rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,85 @@
+# StableHLO Forward Compatibility Testing
+
+## Background
+
+StableHLO needs to verify both forward and backward compatibility. Currently we
+check backward compatibility by versioning and serializing
+`stablehlo_legalize_to_vhlo.mlir`, and need something similar for forward
+compatibility. This document makes the following proposal:
+
+**Proposal:** Add tests to compare bytes of bytecode file serialized at HEAD
+with existing [backward compatibility bytecode test files](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir.bc).
+
+## Requirements
+
+Forward compatibility testing has the following requirements, to ensure that it
+is both useful and practical:
+
+* **R1.** Tests must detect changes in StableHLO that break forward compatibility.
+* **R2.** Tests must detect LLVM revision bumps that break forward compatibility.
+* **R3.** Tests must scale with releases and execute in a reasonable amount
+of time.
+
+## Proposed Design: Statically Test Forward Compatibility
+
+In this design, forward incompatibilities are detected by serializing a file at
+head and comparing it to a known-good file serialized at the previous release.
+This approach operates on the assumption that at HEAD we are able to produce
+a byte-identical serialized artifact, excluding custom header info and debug
+locations.
+
+Detecting forward incompatiblities can be accomplished by adding the following
+`RUN` line to the versioned [`stablehlo_legalize_to_vhlo.0_X_0.mlir`](https://github.com/search?q=repo%3Aopenxla%2Fstablehlo+path%3A**%2Fstablehlo_legalize_to_vhlo.0_*&type=code)
+test files.
+
+```bash
+diff %s.bc <(stablehlo-opt %s | stablehlo-translate --serialize --target=0.X.0)
+```
+
+Passing through `stablehlo-opt` in this example removes debug info. An option
+to `stablehlo-translate` can be added to strip debug info and header info. This
+test should be run in CI on all PRs to prevent merging incompatible changes in
+StableHLO (_R1_) and LLVM revision bumps (_R2_).
```

Comment:
How about rewriting this as follows, to first explain how this approach would work and only then state the assumptions?

```
Detecting forward incompatiblities can be accomplished by adding the following
`RUN` line to the versioned [`stablehlo_legalize_to_vhlo.0_X_0.mlir`](https://github.com/search?q=repo%3Aopenxla%2Fstablehlo+path%3A**%2Fstablehlo_legalize_to_vhlo.0_*&type=code)
test files.

diff %s.bc <(stablehlo-translate --serialize --target=0.X.0)

This test should be run in CI on all PRs to prevent merging incompatible changes 
in StableHLO (_R1_) and LLVM revision bumps (_R2_).

This will not work as is because `bytecode::producer`, `op::location` and 
`block_argument::location` fields may change between when the original bytecode
file has been produced and when the test runs at HEAD. To account for that,
we can strip this information before comparing the payloads.
```